packages/hydrogen-edge-adapter/src/server.js

```diff
@@ -0,0 +1,123 @@
+// Virtual entry point for the app
+import * as remixBuild from '@remix-run/dev/server-build'
+import { createRequestHandler as netlifyCreateRequestHandler } from '@netlify/remix-edge-adapter'
+import { createRequestHandler as hydrogenCreateRequestHandler, getStorefrontHeaders } from '@shopify/remix-oxygen'
+import { createStorefrontClient, storefrontRedirect } from '@shopify/hydrogen'
+
+export async function getHydrogenClient({ env, request, waitUntil, caches, HydrogenSession, getLocaleFromRequest }) {
+  /**
+   * Open a cache instance in the worker and a custom session instance.
+   */
+  if (!env?.SESSION_SECRET) {
+    throw new Error('SESSION_SECRET environment variable is not set')
+  }
+
+  const [cache, session] = await Promise.all([
+    caches ? caches.open('hydrogen') : Promise.resolve(undefined),
+    HydrogenSession.init(request, [env.SESSION_SECRET]),
+  ])
+
+  /**
+   * Create Hydrogen's Storefront client.
+   */
+  const { storefront } = createStorefrontClient({
+    cache,
+    waitUntil,
+    i18n: getLocaleFromRequest(request),
+    publicStorefrontToken: env.PUBLIC_STOREFRONT_API_TOKEN,
+    privateStorefrontToken: env.PRIVATE_STOREFRONT_API_TOKEN,
+    storeDomain: env.PUBLIC_STORE_DOMAIN,
+    storefrontId: env.PUBLIC_STOREFRONT_ID,
+    storefrontHeaders: getStorefrontHeaders(request),
+  })
+
+  return {
+    storefront,
+    session,
+  }
+}
+
+export function createHydrogenHandler(overrides) {
+  return async function (request, env, executionContext) {
+    const waitUntil = (p) => executionContext.waitUntil(p)
+    const { storefront, session } = await getHydrogenClient(
+      Object.assign(
+        {
+          request,
+          env,
+          waitUntil,
+          caches,
```

Comment:
Where is `caches` coming from?

Reply:
`caches` (confusingly) is a global via CacheStorage:
* https://developer.mozilla.org/en-US/docs/Web/API/CacheStorage
* https://developer.mozilla.org/en-US/docs/Web/API/caches

Reply:
It could be clearer if I don’t pass it in though, I’ll make that change.